### PR TITLE
Adding WinHTTP Websocket support

### PR DESCRIPTION
**Pull request by jasonsandlin**

### CODE REVIEW

**Comment by jplafonta** [comment]
> // Copyright (c) Microsoft Corporation

Make sure this builds against win7

---
Refers to: Source/HTTP/WinHttp/winhttp_http_task.cpp:1 in 8553530. [](commit_id = 85535309fa725fb5a1c7d32cf517e55dc300dc91, deletion_comment = False)

File: Source/WebSocket/WinHTTP/winhttp_websocket.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft Corporation
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pch.h"
+
+#if HC_WINHTTP_WEBSOCKETS
+#if !HC_NOWEBSOCKETS
```
Comment (jplafonta) [inline Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: seems a bit odd - I think these should never be defined together right?

File: Source/WebSocket/WinHTTP/winhttp_websocket.cpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+using namespace xbox::httpclient;
+
+
+struct winhttp_websocket_impl : public hc_websocket_impl, public std::enable_shared_from_this<winhttp_websocket_impl>
+{
+public:
+    winhttp_websocket_impl(HCWebsocketHandle hcHandle)
+    {
+        m_hcWebsocketHandle = HCWebSocketDuplicateHandle(hcHandle);
+        // TODO: when to close
```
Comment (jplafonta) [inline Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: >/ TODO: when to close [](start = 9, length = 21)

I think you don't actually want to duplicate the handle here.  iirc this object is owned by the HC_WEBSOCKET, so duplicated the handle here will cause it to never get cleaned up.

Comment (jplafonta) [reply Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: the websocket_impl object's lifetime is always the same as the HC_WEBSOCKET

---
In reply to: [248444615](https://github.com/Microsoft/libHttpClient/pull/361#discussion_r248444615) [](ancestors = 248444615)

File: Source/WebSocket/hcwebsocket.cpp [hunk 3]
```diff
@@ -101,25 +109,71 @@ try
...
+            case XAsyncOp::GetResult:
+                break;
+
+            case XAsyncOp::Cancel:
+                break;
+
+            case XAsyncOp::Cleanup:
+            {
+                websocket_connect_context* context = static_cast<websocket_connect_context*>(data->context);
+                //HCWebSocketCloseHandle(context->websocket); // TODO: jasonsa - fix for WinHTTP
```
Comment (jplafonta) [inline Source/WebSocket/hcwebsocket.cpp]: where does this get duplicated?

File: Source/WebSocket/hcwebsocket.cpp [hunk 4]
```diff
@@ -101,25 +109,71 @@ try
...
+    std::shared_ptr<websocket_connect_context> connectContext = std::make_shared<websocket_connect_context>();
+    connectContext->uri = uri;
+    connectContext->subProtocol = subProtocol;
+    connectContext->websocket = websocket;
+    connectContext->outerAsyncBlock = asyncBlock;
+    websocket_connect_context* rawConnectContext = static_cast<websocket_connect_context*>(shared_ptr_cache::store<websocket_connect_context>(connectContext));
 
-    auto connectFunc = httpSingleton->m_websocketConnectFunc;
-    if (connectFunc != nullptr)
+    HRESULT hr = XAsyncBegin(asyncBlock, rawConnectContext, reinterpret_cast<void*>(HCWebSocketConnectAsync), __FUNCTION__,
```
Comment (jplafonta) [inline Source/WebSocket/hcwebsocket.cpp]: why did you change this such that the connect is kicked off asynchronously?

File: Source/WebSocket/WinHTTP/winhttp_websocket.cpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    _In_ HCWebsocketHandle websocket,
+    _Inout_ XAsyncBlock* asyncBlock,
+    _In_ HCPerformEnv env
+    )
+{
+    assert(env != nullptr);
+
+    websocket->uri = uri;
+    websocket->subProtocol = subProtocol;
+    auto wsppSocket = http_allocate_shared<winhttp_websocket_impl>(websocket);
```
Comment (jplafonta) [inline Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: rename

Comment (jplafonta) [inline Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: seems like we should check that we aren't already connected here? Maybe set websocket->impl to null when we disconnect?  I don't think we are doing this in wspp but should probably be fixed in both places.

File: Source/WebSocket/hcwebsocket.cpp (hunk 4 above)
Comment (jasonsandlin) [reply Source/WebSocket/hcwebsocket.cpp]: Ah, good catch.  This is wrong

File: Source/WebSocket/WinHTTP/winhttp_websocket.cpp (hunk 5 above)
Comment (jasonsandlin) [reply Source/WebSocket/WinHTTP/winhttp_websocket.cpp]: You should be able to disconnect/reconnect without tossing handle so impl != null isn't right logic, but I'll add a TODO to guard against double connect

File: Source/WebSocket/hcwebsocket.cpp (hunk 3 above)
Comment (jasonsandlin) [reply Source/WebSocket/hcwebsocket.cpp]: bug.  it should be above.  the intent is that we keep the HCWebsocketHandle alive while connecting/connected, and let it be destroyed only when its disconnected.  we'll have to chat a bit about this
